decidim-core/app/queries/decidim/metrics/participants_metric_manage.rb

```diff
@@ -0,0 +1,146 @@
+      #
+      #  - ParticipatorySpace (type & ID)
+      def query
+        return @query if @query
+
+        @query = retrieve_participatory_spaces.each_with_object({}) do |participatory_space, _grouped_participants|
+          components = retrieve_components(participatory_space)
+          components.each do |component|
+            # puts "---------- #{component.manifest_name}"
+            # puts send(:"retrieve_participants_for_#{component.manifest_name}", component)
+            send(:"retrieve_participants_for_#{component.manifest_name}", component)
+          end
+          #   related_object = comment.root_commentable
+          #   return grouped_comments unless related_object
+          #
+          #   group_key = generate_group_key(related_object)
+          #   grouped_comments[group_key] ||= { cumulative: 0, quantity: 0 }
+          #   grouped_comments[group_key][:cumulative] += 1
+          #   grouped_comments[group_key][:quantity] += 1 if comment.created_at >= start_time
+          #
+          #   grouped_comments
+        end
+
+        @query
+      end
+
+      # Search for all Participatory Space manifests and then all records available
+      # Limited to ParticipatoryProcesses only
+      def retrieve_participatory_spaces
+        Decidim.participatory_space_manifests.map do |space_manifest|
+          next unless space_manifest.name == :participatory_processes # Temporal limitation
+          space_manifest.participatory_spaces.call(@organization)
+        end.flatten
+      end
+
+      # Search for all components published, within a fixed list of available
+      def retrieve_components(participatory_space)
+        participatory_space.components.published.where(manifest_name: AVAILABLE_COMPONENTS)
+      end
+
+      def retrieve_participants_for_proposals(component)
+        proposals = Decidim::Proposals::Proposal.where(component: component).joins(:component)
+                                                .includes(:votes, :endorsements)
+                                                .except_withdrawn
+
+        votes = Decidim::Proposals::ProposalVote.joins(:proposal).where(proposal: proposals)
+                                                .where("decidim_proposals_proposal_votes.created_at <= ?", end_time)
+        endorsements = Decidim::Proposals::ProposalEndorsement.joins(:proposal).where(proposal: proposals)
+                                                              .where("decidim_proposals_proposal_endorsements.created_at <= ?", end_time)
+        proposals = proposals.where("decidim_proposals_proposals.published_at <= ?", end_time)
+
+        cumulative_users = []
+        cumulative_users |= votes.collect(&:author).select { |u| u.class == Decidim::User }.uniq
+        cumulative_users |= endorsements.collect(&:author).select { |u| u.class == Decidim::User }.uniq
+        cumulative_users |= proposals.collect(&:authors).flatten.select { |u| u.class == Decidim::User }.uniq
```

Comment:
A better strategy could be to collect all the author ids and types and then search for the users. Something like:

```ruby
user_ids = proposals.includes(:coauthorships).where(coauthorships: { decidim_author_type: "Decidim::UserBaseEntity"}.pluck(:decidim_author_id)

# merge user_ids with the ones from votes and endorsements

cumulative = Decidim::User.where(id: user_ids).uniq
```

Reply:
Seems cleaner and readable than mine, thanks!